core/src/main/java/org/infinispan/remoting/inboundhandler/BasePerCacheInboundInvocationHandler.java

```diff
@@ -66,6 +69,7 @@ public void injectDependencies(@ComponentName(REMOTE_COMMAND_EXECUTOR) BlockingT
       this.cancellationService = cancellationService;
       this.stateTransferLock = stateTransferLock;
       this.stateTransferManager = stateTransferManager;
+      this.async = !configuration.clustering().cacheMode().isSynchronous();
```

Comment:
Even in asynchronous caches, some commands will be synchronous (and vice versa). You need to check each command's `DeliverOrder` instead.